### PR TITLE
Add "exports" field to use-sync-external-store's package.json

### PR DESCRIPTION
**Pull request by latin-1**

### CODE REVIEW

**Comment by gaearon** [comment]
doesn't this also need `./`?

**Comment by latin-1** [comment]
Oops, it appears I overlooked something. Updated.

**Comment by latin-1** [comment]
I'm not sure if `"./src/*": "./src/*"` is necessary for this package. Should I include that in the `package.json` as well?

**Comment by sachinraja** [comment]
This would help support React Native and be esm compliant. In https://github.com/tannerlinsley/react-query/issues/3582 RN support was broken because we imported from `use-sync-external-store/shim/index.js` to support esm which does not support directory imports and must use extensions. With this PR we can import from `use-sync-external-store/shim` which would work correctly for esm and RN.

**Comment by TkDodo** [comment]
@gaearon @acdlite I think this PR would fix our issues we have over in react-query with the uSES shim for react native. apollo-client is having the same issue, and they have opted to copy over the uSES code to fix it on their side. I would rather wait for an official fix on your end, which I think this PR provides :)

File: packages/use-sync-external-store/package.json [hunk 1]
```diff
@@ -17,6 +17,15 @@
     "shim/",
     "cjs/"
   ],
+  "exports": {
+    ".": "./index.js",
+    "./with-selector": "./with-selector.js",
+    "./shim": "./shim/index.js",
```
Comment (SimenB) [inline packages/use-sync-external-store/package.json:23]: thoughts on

```suggestion
    "./shim": {
      "react-native": "./shim/index.native.js",
      "default": "./shim/index.js"
    },
```

?

That way all code can just be `use-sync-external-store/shim`.

note that this relies on the bundler actually supporting `exports` and supplying `react-native` condition (which I'm guessing `metro` at least doesn't do), but it seems more future-proof than these (seemingly) somewhat random exports

Comment (latin-1) [reply packages/use-sync-external-store/package.json:23]: https://github.com/facebook/metro/issues/670
Metro doesn't respect `exports` field at all.

Comment (SimenB) [reply packages/use-sync-external-store/package.json:23]: Right, so no issue.

This stanza should still work for any modern bundler, and metro whenever it enters the future

**Comment by gaearon** [comment]
@latin-1 What makes this a breaking change?

**Comment by sachinraja** [comment]
@gaearon You can no longer import with extensions in ESM and some paths have changed:
```
use-sync-external-store/index.js -> use-sync-external-store
use-sync-external-store/shim/index.js -> use-sync-external-store/shim
use-sync-external-store/with-selector.js -> use-sync-external-store/with-selector
```

We may be able to support it by just exporting everything under the same path as before (maybe using glob exports).

**Comment by TkDodo** [comment]
@sachinraja @gaearon @latin-1 what's the status of this PR? If the change are okay, could we release it as a new major version? We are still having troubles with the uSES shim in react-native because of it. Thanks 🙏

**Comment by eps1lon** [comment]
> We are still having troubles with the uSES shim in react-native because of it. Thanks pray

Don't we need https://github.com/facebook/react/pull/24440#discussion_r868956506 for that?

File: packages/use-sync-external-store/package.json (hunk 1 above)
Comment (eps1lon) [reply packages/use-sync-external-store/package.json:23]: There's Repack to consider (basically Webpack for react-native) so we might as well get this in now. Doing it in the future might be considered another breaking change.

File: packages/use-sync-external-store/package.json [hunk 2]
```diff
@@ -17,6 +17,15 @@
...
+    "./shim/index.native": "./shim/index.native.js",
+    "./shim/with-selector": "./shim/with-selector.js",
+    "./src/*": "./src/*",
```
Comment (eps1lon) [inline packages/use-sync-external-store/package.json:26]: Wouldn't this require changing the publish script? The latest version of `uSES` has no `src` folder. What are you currently importing with `use-sync-external-store/src/?`?

File: packages/use-sync-external-store/index.js [hunk 3]
```diff
@@ -9,4 +9,4 @@
 
 'use strict';
 
-export {useSyncExternalStore} from './src/useSyncExternalStore';
```
Comment (eps1lon) [inline packages/use-sync-external-store/index.js:12]: Pretty sure these get re-written automatically and should stay relative imports: https://unpkg.com/browse/use-sync-external-store@1.2.0/index.js

**Comment by eps1lon** [comment]
Alternate that's not exposing non-existent `src` and tested: https://github.com/facebook/react/pull/25231

File: packages/use-sync-external-store/package.json (hunk 2 above)
Comment (latin-1) [reply packages/use-sync-external-store/package.json:26]: It seems that the react package has this inline.
https://github.com/facebook/react/blob/3613284dceac38de4ef5a7c93c4897600ee6923b/packages/react/package.json#L30

**Comment by latin-1** [comment]
Closing in favor of #25231

**Pull request closed without merging**